### ISSUE TITLE
resizing of canvas broken in combination with useWorker

### ISSUE THREAD

**Issue by timurc** [catdad/canvas-confetti#104]
First thanks a lot for this package, made celebrating hitting a milestone in our organisation a lot more fun! https://expedition-grundeinkommen.de/hamburg/

This is what I see in the console when using `useWorker: true`:

```
Uncaught TypeError: canvas.getBoundingClientRect is not a function
```

Only in combination with supplying my own canvas element though.

I just glimpsed at the code and it seems like `setCanvasRectSize` expects a canvas element but is called with an empty object as `resizer` here: https://github.com/catdad/canvas-confetti/blob/master/src/confetti.js#L475

I am not that experienced in forking and local npm package development, that's why I did not offer a pull reqest to fix it and just filed this issue for now. If that is appreciated I would give it a try.

**Comment by catdad** [comment]
Thanks for reaching out! I'll take a look tonight

**Comment by catdad** [comment]
This is fixed and published in v1.1.1

**Comment by timurc** [comment]
Hey thanks, that was quick! It works in FF and Safari, but breaks in Chrome unfortunately. When it scrolls out of view and into the view again, it breaks with 

```
Uncaught DOMException: Failed to set the 'width' property on 'HTMLCanvasElement': Cannot resize canvas after call to transferControlToOffscreen().
```

You can try to reproduce here: 
https://deploy-preview-90--expedition-grundeinkommen.netlify.com/

You can check out how I used it here: 
https://github.com/grundeinkommensbuero/website/blob/5f6fcd80a1ca943b743c62037c8379227903fca0/src/components/Confetti/index.js

I stop the confetti with IntersectionObserver when it's out of view. Since it does work without problems in the old version without worker, I do not think that that's the problem. (And btw. that might be a cool feature to have build-in, maybe pause the animations completely :-) )

**Comment by catdad** [comment]
Thanks for the link to your code, that was helpful. It looks like you are using it in a way that I never intended (but usually works by accident). What's happening in the `useEffect` hook is that every time `isInView` changes, a new instance of `confetti` is initialized, but using the same `canvas` element. Long story short, I never intended to be able to initialize the same canvas multiple times. React's refs make this a little awkward though.

The issue is that when using workers for rendering, you have to relinquish control of the canvas element entirely to the worker, after which, the main thread can no longer use the canvas. I really need to document this, since it can go wrong in many ways. Without workers, it is not a problem, since the canvas can always be manipulated by the main thread. (It also works in browsers other than Chrome because Chrome is the only one that supports canvas rendering inside workers.)

In your case, where you might need to initialize a canvas multiple times (since you don't really know when React might remount the component), I would consider doing something like this:

```javascript
const canvas = useRef(null);
const confetti = useRef({});

useEffect(() => {
  if (!canvas) {
    return;
  }

  if (canvas.current !== confetti.current.canvas) {
    confetti.current.canvas = canvas.current;
    confetti.current.trigger = canvasConfetti.create(canvas.current, { resize: true, useWorker: true });
  }

  // fire confetti however you like now
  confetti.current.trigger({
    count: 1,
    ...
  });
}, [isInView]);
```

Actually, writing this, I noticed another issue in your code. If the component does ever remount, you will only see one confetti be fired since your `requestAnimationFrame` loop depends on `animating = true` but it never actually gets reset. It's probably safe to remove `animating` entirely, since `isInView` already handles pausing and resuming properly. But I also might be misunderstanding what you are trying to do.

Anyway, I'll update the documentation right away, as this is really important to understand about using workers (it's partially why I left if off by default for user canvases). As for preventing initialization with the same canvas more than once, I'd have to think about how to handle that, since all the ideas I have now would result in a memory leak or in modifying the canvas. I did create #106 if you want to track this new bug.

**Comment by timurc** [comment]
Hey, I completely missed that I initiated it too often, that is of course not how it should be. I updated it. Yes, you can initialize things just once per mount with react I just messed up 🙃. I also now add use the `.reset()` when it unmounts. I missed that function because somehow I was looking for a `.destroy()` and could not find it.

https://github.com/grundeinkommensbuero/website/blob/6cdc5abaf8ae3f45133bd68a0991c14f290dc314/src/components/Confetti/index.js

I need to break the loop with the `animating` flag because when you navigate away to another page it would keep running and crash because the canvas is not there anymore because the component unmounted in the meantime. The function that is returned from the `useEffect` hook is called when a component unmounts. It does work now—you can test it by navigating to another page and then back to the start page by clicking the logo. It's SPA with the help of gatsbyjs so it won't do a complete reload.

https://deploy-preview-90--expedition-grundeinkommen.netlify.com

If you could somehow detect that it has been initialized too often, I would have spotted that faster than a hint in the documentation. But just nice to have :-)

**Comment by catdad** [comment]
I published an update (v1.1.2) which now allows initializing the control more than once. There are still issues though, which I may address later, but it mostly works. It is still not something that should be done realistically though. Thanks for pointing out this possible use case.
